src/stream-chat-net-test/PermissionTests.cs

```diff
@@ -0,0 +1,135 @@
+
+namespace StreamChatTests
+{
+    /// <summary>
+    /// Even if we create/update/delete a new role, we still need
+    /// to wait a few seconds for the changes to propagate in the backend.
+    /// For this reason, we wrote just a single test to test all functionalites at once.
+    /// </summary>
+    public class PermissionTests
+    {
+        private readonly IClient _client = TestClientFactory.GetClient();
+        private const string TestPermissionDescription = "Test Permission";
+
+        [OneTimeSetUp]
+        [OneTimeTearDown]
+        public async Task CleanupAsync()
+        {
+            await DeleteCustomRolesAsync();
+            await DeleteCustomPermissonsAsync();
+        }
+
+        private async Task DeleteCustomRolesAsync()
+        {
+            var roleResp = await _client.Permissions.ListRoles();
+            foreach (var role in roleResp.Roles.Where(x => x.Custom))
+            {
+                try
+                {
+                    await _client.Permissions.DeleteRole(role.Name);
+                }
+                catch
+                {
+                    // It throws 'role does not exist' exception which
+                    // is obviously nonsense. So let's ignore it.
+                }
+            }
+        }
+
+        private async Task DeleteCustomPermissonsAsync()
+        {
+            var permResp = await _client.Permissions.ListPermissions();
+            foreach (var perm in permResp.Permissions.Where(x => x.Description == TestPermissionDescription))
+            {
+                try
+                {
+                    await _client.Permissions.DeletePermission(perm.Id);
+                }
+                catch
+                {
+                    // It throws 'not found' exception which
+                    // is obviously nonsense. So let's ignore it.
+                }
+            }
+        }
+
+        [Test]
+        public async Task TestRolesEnd2endAsync()
+        {
+            // Test create
+            var roleResp = await _client.Permissions.CreateRole(Guid.NewGuid().ToString());
+
+            Assert.That(roleResp.Role.Name, Is.Not.Null);
+            Assert.That(roleResp.Role.Custom, Is.True);
+            Assert.That(roleResp.Role.CreatedAt, Is.EqualTo(DateTime.UtcNow).Within(TimeSpan.FromSeconds(10)));
+            Assert.That(roleResp.Role.UpdatedAt, Is.EqualTo(DateTime.UtcNow).Within(TimeSpan.FromSeconds(10)));
+
+            // Test list
+            var rolesResp = await _client.Permissions.ListRoles();
+            Assert.That(rolesResp.Roles, Is.Not.Empty);
+
+            // Test delete
+            try
+            {
+                await _client.Permissions.DeleteRole(roleResp.Role.Name);
+            }
+            catch (StreamChatException ex)
+            {
+                if (ex.Message.Contains("does not exist"))
+                {
+                    // Unfortounatly, the backend is too slow to propagate the role creation
```

Comment:
let's have some kind of retry mechanism in place. Otherwise, these tests aren't useful.

Reply:
I actually implemented it with a retry logic first. After it still didn't succeed in 40 seconds I have removed all of that code. It's just too flaky, and it's unacceptable to wait 1 minute for a test to succeed.

I think for now, it's enough that the API calls itself succeed.

Reply:
This is really too much, will check server side of this